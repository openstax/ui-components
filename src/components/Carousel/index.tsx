import React from "react";
import { 
    CarouselContainer, 
    CarouselOverflow, 
    CarouselWrapper, 
    CarouselItem, 
    StyledArrow,
} from './styles';
import { LeftArrow } from "../svgs/LeftArrow";
import { RightArrow } from "../svgs/RightArrow";

export interface CarouselProps {
    children: React.ReactNode[];
}

export const Carousel = ({ children }: CarouselProps) => {
    const [currentIndex, setCurrentIndex] = React.useState(0);
    const [dragStartX, setDragStartX] = React.useState<number | null>(null);
    const [dragX, setDragX] = React.useState(0);
    const [dragging, setDragging] = React.useState(false);
    const wrapperRef = React.useRef<HTMLDivElement>(null);
    const [isLeftArrowDisabled, setIsLeftArrowDisabled] = React.useState(false);
    const [isRightArrowDisabled, setIsRightArrowDisabled] = React.useState(false);
    const scrollSpeed = 100;

    const checkIfRightArrowShouldBeDisabled = React.useCallback(() => {
        if (wrapperRef.current) {
            const isDisabled = (currentIndex >= ((wrapperRef.current.scrollWidth || 0) - (wrapperRef.current.clientWidth || 0)));
            setIsRightArrowDisabled(isDisabled);
        }
    }, [currentIndex, wrapperRef]);

    const checkIfLeftArrowShouldBeDisabled = React.useCallback(() => {
        setIsLeftArrowDisabled((currentIndex) === 0);
    }, [currentIndex]);

    React.useEffect(() => {
        if (wrapperRef.current) {
            checkIfLeftArrowShouldBeDisabled();
            checkIfRightArrowShouldBeDisabled();

            window.addEventListener('resize', checkIfLeftArrowShouldBeDisabled);
            window.addEventListener('resize', checkIfRightArrowShouldBeDisabled);
        }
        return () => {
            window.removeEventListener('resize', checkIfLeftArrowShouldBeDisabled);
            window.removeEventListener('resize', checkIfRightArrowShouldBeDisabled);
        };
    }, [currentIndex, wrapperRef, checkIfLeftArrowShouldBeDisabled, checkIfRightArrowShouldBeDisabled]);

    const handlePrev = () => {
        if (wrapperRef.current && !isLeftArrowDisabled) {
            const itemWidth = (wrapperRef.current.clientWidth / children.length);
            setCurrentIndex((prevIndex) => Math.max(prevIndex - itemWidth - scrollSpeed, 0));
        }
    };

    const handleNext = () => {
        if (wrapperRef.current && !isRightArrowDisabled) {
            const itemWidth = (wrapperRef.current.clientWidth / children.length);
            const maxIndex = (wrapperRef.current.scrollWidth - wrapperRef.current.clientWidth);
            setCurrentIndex((prevIndex) => Math.min(prevIndex + itemWidth + scrollSpeed, maxIndex));
        }
    };

    const handleMouseDown = (e: React.MouseEvent) => {
        setDragStartX(e.clientX);
        setDragging(true); // Allow dragging when move cursor
    };

    const handleMouseMove = (e: React.MouseEvent) => {
        if (!dragging || dragStartX === null) return;
        const dragDistance = (e.clientX - dragStartX) / 10;
        setDragX(dragDistance);
        if (wrapperRef.current) {
            wrapperRef.current.style.transition = 'none';
            wrapperRef.current.style.transform = `translateX(calc(-${(currentIndex / 10)}rem + ${dragDistance}rem))`;
        }
    };

    const handleMouseUp = () => {
        if (!dragging) return;
        setDragging(false);
        setDragStartX(null);
        if (wrapperRef.current) {
            const newIndex = currentIndex - (dragX * 10); // *10 convert from rem to px 
            const maxIndex = (wrapperRef.current.scrollWidth - wrapperRef.current.clientWidth);
            const clampedIndex = Math.max(0, Math.min(newIndex, maxIndex));
            setCurrentIndex(clampedIndex);
            setDragX(0);
            wrapperRef.current.style.transition = 'transform 0.3s ease-in-out';
            wrapperRef.current.style.transform = `translateX(-${(clampedIndex / 10)}rem)`;
        }
    };

    return (
        <CarouselContainer>
            <StyledArrow
                onClick={handlePrev}
                className="left-arrow"
                disabled={isLeftArrowDisabled}
            >
                <LeftArrow />
            </StyledArrow>
            <CarouselOverflow
                onPointerDown={handleMouseDown}
                onPointerMove={handleMouseMove}
                onPointerUp={handleMouseUp}
                onPointerLeave={handleMouseUp}
            >
                <CarouselWrapper
                    ref={wrapperRef}
                    // Updates the position of the carousel when the currentIndex changes, avoiding problems when the user uses the keyboard to navigate
                    style={{ transform: `translateX(calc(-${(currentIndex / 10)}rem + ${dragX}rem))` }}
                >
                    {children.map((child) => {
                        const key = (child as React.ReactElement).key;
                        return (
                            <CarouselItem
                                key={key ?? null}
                            >
                                {child}
                            </CarouselItem>);
                    })}
                </CarouselWrapper>
            </CarouselOverflow>
            <StyledArrow
                onClick={handleNext}
                className="right-arrow"
                disabled={isRightArrowDisabled} >
                    <RightArrow />
            </StyledArrow>
        </CarouselContainer>

    );
};