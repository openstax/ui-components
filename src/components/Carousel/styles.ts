import styled from 'styled-components';
import { LeftArrow } from "../svgs/LeftArrow";
import { RightArrow } from "../svgs/RightArrow";

export const CarouselContainer = styled.div<{customWidth?: string}>`
  position: relative;
  width: ${({customWidth}) =>  customWidth ? customWidth : '30rem'};
  margin: 0 4rem;
`;

export const CarouselOverflow = styled.div`
  position: inherit;
  overflow: hidden;
`;

export const CarouselWrapper = styled.div`
  display: flex;
  width: auto;
  transition: transform 0.3s ease-in-out;
`;

export const CarouselItem = styled.div`
  flex: 0 0 auto;
  margin-right: 1rem;
`;

export const StyledLeftArrow = styled(LeftArrow)<{ disabled: boolean }>`
  position: absolute;
  top: 50%;
  left: -2rem;
  transform: translateY(-50%);
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};
`;

export const StyledRightArrow = styled(RightArrow)<{ disabled: boolean }>`
  position: absolute;
  top: 50%;
  right: -2rem;
  transform: translateY(-50%);
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};
`;