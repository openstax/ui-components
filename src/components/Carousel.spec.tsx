import { Carousel } from './Carousel/index';
import renderer from 'react-test-renderer';

describe('Carousel', () => {

    const childrenListWithKeys = [
        <button key='1'>Slide 1</button>,
        <button key='2'>Slide 2</button>,
        <button key='3'>Slide 3</button>,
        <button key='4'>Slide 4</button>,
        <button key='5'>Slide 5</button>,
    ];

    it('matches snapshot', () => {
        const tree = renderer.create(
            <Carousel>{childrenListWithKeys}</Carousel>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
