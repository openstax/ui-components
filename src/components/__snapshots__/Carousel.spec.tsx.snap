// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel matches snapshot 1`] = `
<div
  className="sc-bczRLJ UjSde"
>
  <svg
    className="sc-eCYdqJ hCWvEb"
    disabled={false}
    fill="none"
    height="15"
    onClick={[Function]}
    viewBox="0 0 8 10"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1.1266 5.38862L5.57702 9.83902C5.79166 10.0537 6.13964 10.0537 6.35426 9.83902L6.87333 9.31995C7.0876 9.10568 7.08801 8.75841 6.87424 8.54363L3.34721 4.99999L6.87424 1.45637C7.08801 1.24159 7.0876 0.89432 6.87333 0.680047L6.35426 0.160979C6.13962 -0.0536598 5.79164 -0.0536598 5.57702 0.160979L1.12662 4.61138C0.911981 4.826 0.911981 5.17398 1.1266 5.38862Z"
      fill="#959595"
    />
  </svg>
  <div
    className="sc-gsnTZi jJYWnI"
    onPointerDown={[Function]}
    onPointerLeave={[Function]}
    onPointerMove={[Function]}
    onPointerUp={[Function]}
  >
    <div
      className="sc-dkzDqf gzhsGE"
      style={
        Object {
          "transform": "translateX(calc(-0rem + 0rem))",
        }
      }
    >
      <div
        className="sc-hKMtZM dhpram"
      >
        <button>
          Slide 1
        </button>
      </div>
      <div
        className="sc-hKMtZM dhpram"
      >
        <button>
          Slide 2
        </button>
      </div>
      <div
        className="sc-hKMtZM dhpram"
      >
        <button>
          Slide 3
        </button>
      </div>
      <div
        className="sc-hKMtZM dhpram"
      >
        <button>
          Slide 4
        </button>
      </div>
      <div
        className="sc-hKMtZM dhpram"
      >
        <button>
          Slide 5
        </button>
      </div>
    </div>
  </div>
  <svg
    className="sc-jSMfEi cBydFW"
    disabled={false}
    fill="none"
    height="15"
    onClick={[Function]}
    viewBox="0 0 8 10"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M6.87337 5.38862L2.42295 9.83902C2.20831 10.0537 1.86033 10.0537 1.64571 9.83902L1.12664 9.31995C0.912369 9.10568 0.911957 8.75841 1.12573 8.54363L4.65276 4.99999L1.12573 1.45637C0.911957 1.24159 0.912369 0.89432 1.12664 0.680047L1.64571 0.160979C1.86035 -0.0536598 2.20833 -0.0536598 2.42295 0.160979L6.87335 4.61138C7.08799 4.826 7.08799 5.17398 6.87337 5.38862Z"
      fill="#959595"
    />
  </svg>
</div>
`;
